frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2835,31 +2874,52 @@ gatherAndFilterCandidatesForwarding(Context* context,
 
   // Resolve the forwarding expression's types & decide if we
   // want to consider forwarding.
-  const ResolvedFields* forwards = nullptr;
+  ResolvedFields forwards;
   UniqueString name = ci.name();
   if (name == USTR("init") || name == USTR("init=") || name == USTR("deinit")) {
     // these are exempt from forwarding
   } else if (auto ct = receiverType->getCompositeType()) {
-    auto useDefaults = DefaultsPolicy::USE_DEFAULTS;
-    const ResolvedFields& fields = fieldsForTypeDecl(context, ct,
-                                                     useDefaults);
-    if (fields.numForwards() > 0) {
-      // and check for cycles
-      bool cycleFound = emitErrorForForwardingCycles(context, ct);
-      if (cycleFound == false) {
-        forwards = &fields;
+    // Possible recursion here when resolving a function call in a forwarding
+    // statement:
+    //     record R { forwarding foo(); }
+    // We need to try resolving 'foo()' as a method on 'R', which eventually
+    // leads us back to this path here.
+    //
+    // By skipping the 'resolveForwardingExprs' check below, we effectively
+    // prevent forwarding statements from containing expressions that
+    // themselves require forwarding. For example, if you had a couple of
+    // forwarding statements like:
+    //     forwarding b;
+    //     forwarding bar();
+    // The 'isQueryRunning' check below would prevent resolving a method
+    // 'bar()' on 'b'.
+    //
+    if (!context->isQueryRunning(resolveForwardingExprs,
```

Comment:
Sometimes it's hard to find a better way than `isQueryRunning` but in this case, I think it might be better to entirely disable forwarding when resolving all forwarding expressions (e.g. `bar()` in your comment above). To do that, I think you would need `resolveForwardingExprs` to communicate to the Resolver it is working with that it should skip the consideration of `forwarding` expressions. We could add something to DefaultsPolicy for this purpose (and, potentially rename it).

Reply:
Would that require us threading some kind of "ignore forwarding" bool through resolveCall all the way to resolveForwardingExprs?

Reply:
yes, it would. At present, resolveCall isn't actually a query, so adding another argument does not seem particularly problematic to me, in terms of performance/reuse, at least.

Reply:
I suppose you might also be able to have the resolveCall notice that the uAST it is resolving is a forwarding expression and then skip resolveForwardingExprs.